### PR TITLE
New controls

### PR DESCRIPTION
**Pull request by hbons**
Implementation of the new Cocoa toolkit: https://www.figma.com/file/yWyGwz8ihDQyqh9d9S4LLE/master-FigmaSharp-Toolkit-Cocoa-Dark

Notable changes:
- Use View() to wrap controls for now
- Add NSColor support for labels and NSBox
- Labels support font weights
- Add a bunch of new controls
- New base CocoaControl and rely less on extension methods
- CocoaHelpers and CocoaCodeHelpers
- Implement code generation for everything

### CODE REVIEW

**Comment by netonjm** [comment]
Branch seems to be conflicts, could you take a look?

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
+				}
+
+				return builder.ToString();
+			}
+			return string.Empty;
+		}
+
+
+
+		protected NSControlSize GetNSControlSize(NativeControlVariant controlVariant)
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:130]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        {
+			if (controlVariant == NativeControlVariant.Small)
+				return NSControlSize.Small;
+
+			return NSControlSize.Regular;
+        }
+
+
+		protected string GetNSFontSizeName(NativeControlVariant controlVariant)
+		{
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:140]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 3]
```diff
@@ -0,0 +1,348 @@
...
+			if (controlVariant == NativeControlVariant.Small)
+				return $"{ typeof(NSFont) }.{ nameof(NSFont.SmallSystemFontSize) }";
+
+			return $"{ typeof(NSFont) }.{ nameof(NSFont.SystemFontSize) }";
+		}
+
+		protected nfloat GetNSFontSize(NativeControlVariant controlVariant)
+		{
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:148]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 4]
```diff
@@ -0,0 +1,348 @@
...
+			if (controlVariant == NativeControlVariant.Small)
+				return NSFont.SmallSystemFontSize;
+
+			return NSFont.SystemFontSize;
+		}
+
+
+		protected NSFont GetNSFont(NativeControlVariant controlVariant)
+		{
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:157]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 5]
```diff
@@ -0,0 +1,348 @@
...
+			if (controlVariant == NativeControlVariant.Small)
+				return NSFont.SystemFontOfSize(NSFont.SmallSystemFontSize);
+
+			return NSFont.SystemFontOfSize(NSFont.SystemFontSize);
+		}
+
+		protected NSFont GetNSFont(NativeControlVariant controlVariant, FigmaText text)
+		{
+			var fontWeight = GetNSFontWeight(text);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:166]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse, please back again to the corresponding class

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+				else
+					return NSFont.SystemFontOfSize(NSFont.SmallSystemFontSize, fontWeight);
+			}
+
+			return NSFont.SystemFontOfSize(GetNSFontSize(controlVariant), fontWeight);
+		}
+
+		protected string GetNSFontName(NativeControlVariant controlVariant)
+		{
+			if (controlVariant == NativeControlVariant.Small)
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:189]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+				return $"{ typeof(NSFont) }.{ nameof(NSFont.SystemFontOfSize) }({ GetNSFontSizeName(controlVariant) })";
+
+			return $"{ typeof(NSFont) }.{ nameof(NSFont.SystemFontOfSize) }({ GetNSFontSizeName(controlVariant) })";
+		}
+
+		protected string GetNSFontName(NativeControlVariant controlVariant, FigmaText text, bool withWeight = true)
+		{
+			var fontWeight = GetNSFontWeight(text);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:197]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 8]
```diff
@@ -0,0 +1,348 @@
...
+			    return $"{ typeof(NSFont) }.{ nameof(NSFont.SystemFontOfSize) }({ GetNSFontSizeName(controlVariant) }, { GetNSFontWeightName(text) })";
+            else
+				return $"{ typeof(NSFont) }.{ nameof(NSFont.SystemFontOfSize) }({ GetNSFontSizeName(controlVariant) })";
+
+		}
+
+
+		public nfloat GetNSFontWeight(FigmaText text)
+		{
+			string fontName = text?.style?.fontPostScriptName;
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:225]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+			}
+
+			// The default macOS font is of medium weight
+			return NSFontWeight.Medium;
+		}
+
+		protected string GetNSFontWeightName(FigmaText text)
+		{
+            // The default macOS font is of medium weight
+            string weight = nameof(NSFontWeight.Medium);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:261]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+					weight = nameof(NSFontWeight.UltraLight);
+			}
+
+			return $"{ typeof(NSFontWeight) }.{ weight }";
+		}
+
+
+		protected NSTextAlignment GetNSTextAlignment(FigmaText text)
+		{
+			FigmaTypeStyle style = text.style;
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:297]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 11]
```diff
@@ -0,0 +1,348 @@
...
+
+		public string GetNSTextAlignmentName(FigmaText text)
+		{
+			return $"{nameof(NSTextAlignment)}.{GetNSTextAlignment(text)}";
+		}
+
+
+		public NSColor GetNSColor(string colorStyleName)
+		{
+			return themeColors.FirstOrDefault(c => c.styleName == colorStyleName).themeColor;
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:316]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+		}
+
+		public string GetNSColorName(string colorStyleName)
+		{
+			return $"{nameof(NSColor)}.{themeColors.FirstOrDefault(c => c.styleName == colorStyleName).themeColorName}";
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:321]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+				return NSTextAlignment.Right;
+
+			if (style.textAlignHorizontal == "CENTER")
+				return NSTextAlignment.Center;
+
+			return NSTextAlignment.Left;
+		}
+
+		public string GetNSTextAlignmentName(FigmaText text)
+		{
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:309]: This is a conversion method and it should be moved to a Helper or an Extension class for reuse

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+			return $"{nameof(NSColor)}.{themeColors.FirstOrDefault(c => c.styleName == colorStyleName).themeColorName}";
+		}
+
+
+		// Not all controls in Cocoa can set every a11y property,
+		// this lets converters indicate which properties they can provide
+		public virtual bool CanSetAccessibilityLabel => true;
+		public virtual bool CanSetAccessibilityHelp => true;
+		public virtual bool CanSetAccessibilityRole => true;
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:329]: 👍

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.ThemeColors.cs [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+
+using System.Collections.Generic;
+using AppKit;
+
+namespace FigmaSharp.NativeControls.Cocoa
+{
+	public abstract partial class CocoaConverter
+	{
+		static readonly IReadOnlyList<(string styleName, NSColor themeColor, string themeColorName)> themeColors
+                           = new List<(string styleName, NSColor themeColor, string themeColorName)>
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.ThemeColors.cs:34]: This dictionary of colors should not be here, it must be in some kind of static Resources or some ColorService

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs [hunk 16]
```diff
@@ -40,95 +37,132 @@
 
 namespace FigmaSharp.NativeControls.Cocoa
 {
-    public class TextFieldConverter : FigmaNativeControlConverter
+    public class TextFieldConverter : CocoaConverter
 	{
 		public override Type GetControlType(FigmaNode currentNode)
 		{
+			FigmaNode optionsGroup = currentNode.GetChildren()
+                .FirstOrDefault(s => s.name == "!options" && s.visible);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs:45]: options should be replaced as a constant

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs [hunk 17]
```diff
@@ -40,95 +37,132 @@
...
+			var frame = (FigmaFrame) currentNode;
+			frame.TryGetNativeControlType(out var controlType);
+			frame.TryGetNativeControlVariant(out var controlVariant);
+
+
+			if (controlType == NativeControlType.SearchField)
+				textField = new NSSearchField();
+
+
+			FigmaNode optionsGroup = frame.children.FirstOrDefault(s => s.name == "!options" && s.visible);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs:85]: options should be replaced as a constant

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs [hunk 18]
```diff
@@ -40,95 +37,132 @@
...
-			instance.TryGetNativeControlComponentType(out var controlType);
-			switch (controlType)
-			{
-				case NativeControlComponentType.TextFieldStandard:
-				case NativeControlComponentType.TextFieldStandardDark:
-					builder.WriteEquality(currentNode.Name, nameof(NSButton.Font),
-						CodeGenerationHelpers.Font.SystemFontOfSize(CodeGenerationHelpers.Font.SystemFontSize));
-					break;
-			}
+			FigmaNode optionsGroup = frame.children.FirstOrDefault(s => s.name == "!options" && s.visible);
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextFieldConverter.cs:136]: options should be replaced as a constant

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextViewConverter.cs [hunk 19]
```diff
@@ -40,115 +37,108 @@
...
-			{
-				case NativeControlComponentType.TextViewSmall:
-				case NativeControlComponentType.TextViewSmallDark:
-					textView.Font = NSFont.SystemFontOfSize(NSFont.SmallSystemFontSize);
-					break;
-			}
+			frame.TryGetNativeControlVariant(out var controlVariant);
 
-			return view;
+            if (controlVariant == NativeControlVariant.Small)
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/Converters/TextViewConverter.cs:82]: style

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/FigmaSharp.NativeControls.Cocoa.csproj [hunk 20]
```diff
@@ -119,4 +125,14 @@
     </ProjectReference>
   </ItemGroup>
   <Import Project="$(MSBuildExtensionsPath)\Xamarin\Mac\Xamarin.Mac.CSharp.targets" />
+  <ProjectExtensions>
+    <MonoDevelop>
+      <Properties>
+        <Policies>
+          <TextStylePolicy TabWidth="4" TabsToSpaces="False" IndentWidth="4" RemoveTrailingWhitespace="True" NoTabsAfterNonTabs="False" EolMarker="Native" FileWidth="80" scope="text/x-csharp" />
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/FigmaSharp.NativeControls.Cocoa.csproj:132]: 👍

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.ComponentStrings.cs [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+// USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Collections.Generic;
+using AppKit;
+
+namespace FigmaSharp.NativeControls.Cocoa
+{
+	public abstract partial class CocoaConverter
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.ComponentStrings.cs:31]: why we need partial classes in a CocoaConverer?

File: FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs [hunk 22]
```diff
@@ -1,53 +1,60 @@
...
+
 using FigmaSharp.Cocoa;
 using FigmaSharp.Models;
 using FigmaSharp.Services;
 using FigmaSharp.Views;
 
 namespace FigmaSharp.NativeControls.Cocoa
 {
-	public abstract class FigmaNativeControlConverter : FigmaViewConverter
+	public abstract partial class CocoaConverter : FigmaViewConverter
```
Comment (netonjm) [inline FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs:38]: why we need partial classes in a CocoaConverer?

Comment (hbons) [reply FigmaSharp.NativeControls/FigmaSharp.NativeControls.Cocoa/CocoaConverter.cs]: Colours were in the same class, but a different file. Fixed now.

**Comment by netonjm** [comment]
Merged! Good job! Please next time more atomic PR's to make the revision easily.

**Comment by hbons** [comment]
🎉 thanks! yes will do.

**Pull request merged**